packages/shared/src/components/onboarding/FilterOnboardingV4.tsx

```diff
@@ -0,0 +1,175 @@
+import React, { ReactElement, useMemo } from 'react';
+import classNames from 'classnames';
+import { useMutation, useQuery, useQueryClient } from 'react-query';
+import request from 'graphql-request';
+import useFeedSettings from '../../hooks/useFeedSettings';
+import { RequestKey, generateQueryKey } from '../../lib/query';
+import {
+  GET_ONBOARDING_TAGS_QUERY,
+  GET_RECOMMENDED_TAGS_QUERY,
+  Tag,
+  TagsData,
+} from '../../graphql/feedSettings';
+import { graphqlUrl } from '../../lib/config';
+import { disabledRefetch } from '../../lib/func';
+import { Button } from '../buttons/Button';
+import { AlertColor, AlertDot } from '../AlertDot';
+import { SearchField } from '../fields/SearchField';
+import useDebounce from '../../hooks/useDebounce';
+import { useTagSearch } from '../../hooks';
+import type { FilterOnboardingProps } from './FilterOnboarding';
+import useTagAndSource from '../../hooks/useTagAndSource';
+import { Origin } from '../../lib/analytics';
+
+type OnSelectTagProps = {
+  tag: Tag;
+};
+
+const tagsSelector = (data: TagsData) => data?.tags || [];
+
+export function FilterOnboardingV4({
+  className,
+}: FilterOnboardingProps): ReactElement {
+  const queryClient = useQueryClient();
+
+  const { feedSettings } = useFeedSettings();
+  const selectedTags = useMemo(() => {
+    return new Set(feedSettings?.includeTags || []);
+  }, [feedSettings?.includeTags]);
+  const { onFollowTags, onUnfollowTags } = useTagAndSource({
+    origin: Origin.TagsFilter,
+  });
+
+  const [refetchFeed] = useDebounce(() => {
+    const feedQueryKey = [RequestKey.FeedPreview];
+    queryClient.cancelQueries(feedQueryKey);
+    queryClient.invalidateQueries(feedQueryKey);
+  }, 1000);
+
+  const onboardingTagsQueryKey = generateQueryKey(
+    RequestKey.Tags,
+    undefined,
+    'onboardingTags',
+  );
+  const { data: onboardingTags } = useQuery(
+    onboardingTagsQueryKey,
+    async () => {
+      const result = await request<{
+        onboardingTags: TagsData;
+      }>(graphqlUrl, GET_ONBOARDING_TAGS_QUERY, {});
+
+      return result.onboardingTags;
+    },
+    {
+      ...disabledRefetch,
+      staleTime: Infinity,
+      select: tagsSelector,
+    },
+  );
+  const excludedTags = useMemo(() => {
+    if (!onboardingTags) {
+      return [];
+    }
+
+    return [...onboardingTags.map((item) => item.name)];
+  }, [onboardingTags]);
+
+  const [searchQuery, setSearchQuery] = React.useState<string>();
+  const [onSearch] = useDebounce(setSearchQuery, 200);
+
+  const { data: searchResult, isLoading: isSearchLoading } = useTagSearch({
+    value: searchQuery,
+  });
+  const searchTags = searchResult?.searchTags.tags || [];
+
+  const { mutate: recommendTags, data: recommendedTags } = useMutation(
```

Comment:
Really not urgent, but wondering if we could move this to a custom hook perhaps and above fetch query, just so we can potentially re-use this otherwise and clean the logic here a little bit.
As mentioned: not urgent at all.

Reply:
I think keeping it here is fine until concluding experiment. If it wins then we can move to hook and also add tests.

Reply:
Sure that's fine with me :)